salesagility/SuiteCRM
# Massive duplicate and update records

## Description




This new functionality adds the possibility of simultaneously duplicating and updating a set of records selected through the list view of any module.

![mass-duplicate-and-update](https://user-images.githubusercontent.com/61022311/209692482-2a2c0ae2-5bbb-4328-9135-7d783fcb71eb.gif)


New functionality features:
- Added a link in the list view actions: _Duplicate and Mass Update_.
- Using Javascript, pressing this button adds the necessary logic for the submitted form to behave in Mass Duplicate and Update.
- In the `handleMassUpdate()` function (in _include/MassUpdate.php_) we add code that handles the mass updating of records as if they were new records, creating a new id and removing the fetched_row property in the bean.
- When records are duplicated using this new procedure, the relationships of the source record are created in the new one, in the same way that occurs in individual duplication. This affects fields of type relate, flex_relate, and “1:n” relationships where the duplicate record is on the “n” side of the relationship.
- The _Users_ module is excluded from this functionality, due to its own limitations, such as the logical impossibility of having repeated usernames. 

## Motivation and Context

This functionality has been seen as interesting and useful in different contexts where it is necessary to simultaneously duplicate and apply some modification to a set of records, in situations where it is not efficient to use the individual duplication of a record. Any of these situations could be:

- The massive duplication of similar invoices in a new period, where the amount must be modified equally for all records.
- The massive duplication of project tasks, directly modifying the new project to which the tasks are assigned.
- The massive duplication of products from an online store.
- In the context of custom modules used to manage registrations for events that must be renewed each course.
- …

## How To Test This

1) Select one or more records in the list view of any module (except Users)
2) In Bulk Actions, select "Duplicate & Mass update
3) In the Mass Update panel, modify the fields as necessary
4) Click on the "Duplicate and update" button and check that new records have been created and that they include the indicated changes.


## Types of changes


### Final checklist
